crates/wast/src/component/binary.rs

```diff
@@ -0,0 +1,612 @@
+                args.encode(e);
+            }
+            InstanceKind::Component { component, args } => {
+                e.push(0x00);
+                e.push(0x01);
+                component.idx.encode(e);
+                args.encode(e);
+            }
+            InstanceKind::BundleOfComponentExports { args } => {
+                e.push(0x01);
+                args.encode(e);
+            }
+            InstanceKind::BundleOfExports { args } => {
+                e.push(0x02);
+                args.encode(e);
+            }
+            InstanceKind::Import { .. } => unreachable!("should be removed during expansion"),
+        }
+    }
+}
+
+impl Encode for NamedModuleArg<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        self.name.encode(e);
+        e.push(0x02);
+        self.arg.encode(e);
+    }
+}
+
+impl Encode for ModuleArg<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        match self {
+            ModuleArg::Def(def) => {
+                def.idx.encode(e);
+            }
+            ModuleArg::BundleOfExports(..) => {
+                unreachable!("should be expanded already")
+            }
+        }
+    }
+}
+
+impl Encode for NamedComponentArg<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        self.name.encode(e);
+        self.arg.encode(e);
+    }
+}
+
+impl Encode for ComponentArg<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        match self {
+            ComponentArg::Def(def) => {
+                match def.kind {
+                    DefTypeKind::Module => e.push(0x00),
+                    DefTypeKind::Component => e.push(0x01),
+                    DefTypeKind::Instance => e.push(0x02),
+                    DefTypeKind::Func => e.push(0x03),
+                    DefTypeKind::Value => e.push(0x04),
+                }
+                def.idx.encode(e);
+            }
+            ComponentArg::BundleOfExports(..) => {
+                unreachable!("should be expanded already")
+            }
+        }
+    }
+}
+
+impl Encode for Start<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        self.func.encode(e);
+        self.args.encode(e);
+    }
+}
+
+impl Encode for Alias<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        match self.target {
+            AliasTarget::Export { instance, export } => {
+                match self.kind {
+                    AliasKind::Module => {
+                        e.push(0x00);
+                        e.push(0x00);
+                    }
+                    AliasKind::Component => {
+                        e.push(0x00);
+                        e.push(0x01);
+                    }
+                    AliasKind::Instance => {
+                        e.push(0x00);
+                        e.push(0x02);
+                    }
+                    AliasKind::Value => {
+                        e.push(0x00);
+                        e.push(0x04);
+                    }
+                    AliasKind::ExportKind(export_kind) => {
```

Comment:
Yeah I think this will all need refactoring with the new index space splits, and I'm hoping that the index space split will lend a natural structure to the AST here once we implement it. There's a whole mess of test cases that should be passing but are failing tagged with #588 which I hope to get back to once we redo the binary format.